src/encoding.ts

```diff
@@ -0,0 +1,333 @@
+				}
+				else if (encodedChar < 0x200000) {
+					buffer.push(((encodedChar >> 0x12) & 0x07) | 0xF0);
+					buffer.push(((encodedChar >> 0x0C) & 0x3F) | 0x80);
+					buffer.push(((encodedChar >> 0x06) & 0x3F) | 0x80);
+				}
+				buffer.push((encodedChar & 0x3F) | 0x80);
+			}
+		}
+
+		return buffer;
+	},
+
+	/**
+	 * Decode a Utf8 encoded byte buffer into a string
+	 *
+	 * @param data
+	 * The byte buffer to be decoded
+	 */
+	decode(data: ByteBuffer): string {
+		if (!validateDecodingArgs(data)) {
+			return '';
+		}
+
+		let decoded = '';
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			let byte1 = data[i] & 0xFF;
+
+			if ((byte1 & 0x80) === 0) {
+				decoded += decodeUtf8EncodedCodePoint(byte1);
+			}
+			else if ((byte1 & 0xE0) === 0xC0) {
+				let byte2 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte2);
+				byte2 = byte2 & 0x3F;
+				let encodedByte = ((byte1 & 0x1F) << 0x06) | byte2;
+				decoded += decodeUtf8EncodedCodePoint(encodedByte, [0x80, Infinity]);
+			}
+			else if ((byte1 & 0xF0) === 0xE0) {
+				let byte2 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte2);
+				byte2 = byte2 & 0x3F;
+
+				let byte3 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte3);
+				byte3 = byte3 & 0x3F;
+
+				let encodedByte = ((byte1 & 0x1F) << 0x0C) | (byte2 << 0x06) | byte3;
+				decoded += decodeUtf8EncodedCodePoint(encodedByte, [0x0800, Infinity], true);
+			}
+			else if ((byte1 & 0xF8) === 0xF0) {
+				let byte2 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte2);
+				byte2 = byte2 & 0x3F;
+
+				let byte3 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte3);
+				byte3 = byte3 & 0x3F;
+
+				let byte4 = data[++i] & 0xFF;
+				validateUtf8EncodedCodePoint(byte4);
+				byte4 = byte4 & 0x3F;
+
+				let encodedByte = ((byte1 & 0x1F) << 0x0C) | (byte2 << 0x0C) | (byte3 << 0x06) | byte4;
+				decoded += decodeUtf8EncodedCodePoint(encodedByte, [0x010000, 0x10FFFF]);
+			}
+			else {
+				validateUtf8EncodedCodePoint(byte1);
+			}
+		}
+
+		return decoded;
+	}
+}
+
+/**
+ * Hex codec that provides facilities for encoding a string
+ * into an Hex encoded byte buffer as well as decoding an Hex encoded byte buffer
+ * into a string.
+ */
+export const hex: Codec = {
+	/**
+	 * Encode a string into an Hex byte buffer
+	 *
+	 * @param data
+	 * A Hex encoded string
+	 */
+	encode(data: string = ''): number[] {
+		let buffer: number[] = [];
+
+		for (let i = 0, length = data.length; i < length; i += 2) {
+			let encodedChar = parseInt(data.substr(i, 2), 16);
+
+			buffer.push(encodedChar);
+		}
+
+		return buffer;
+	},
+
+	/**
+	 * Decode a Hex encoded byte buffer into a string
+	 *
+	 * @param data
+	 * The byte buffer to be decoded
+	 */
+	decode(data: ByteBuffer): string {
+		if (!validateDecodingArgs(data)) {
+			return '';
+		}
+
+		let decoded = '';
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			decoded += parseInt('' + data[i], 10).toString(16).toUpperCase();
```

Comment:
I was using `parseInt` because it was originally returning a collection of `numbers`. Now it is returning a string so `parseInt` doesn't make sense. Good catch!